File: ios/InfraAgent/Models/UserContext.swift
import Foundation

struct UserContext: Codable {
    var profile: PersonalProfileContext
    var plan: WorkLearningPlanContext
    var preferences: UserPreferenceContext
    var materials: [UserMaterialContext]
}

struct PersonalProfileContext: Codable {
    var id: String
    var displayName: String
    var goal: String
    var backgroundSummary: String
    var currentStage: String
    var constraints: [String]
    var updatedAt: Date

    enum CodingKeys: String, CodingKey {
        case id
        case displayName = "display_name"
        case goal
        case backgroundSummary = "background_summary"
        case currentStage = "current_stage"
        case constraints
        case updatedAt = "updated_at"
    }
}

struct WorkLearningPlanContext: Codable {
    var id: String
    var longTermGoal: String
    var targetCycle: String
    var fullCyclePlan: [String]
    var weeklyFocus: String
    var activeTasks: [String]
    var nextAction: String
    var trackingKeywords: [String]
    var updatedAt: Date

    enum CodingKeys: String, CodingKey {
        case id
        case longTermGoal = "long_term_goal"
        case targetCycle = "target_cycle"
        case fullCyclePlan = "full_cycle_plan"
        case weeklyFocus = "weekly_focus"
        case activeTasks = "active_tasks"
        case nextAction = "next_action"
        case trackingKeywords = "tracking_keywords"
        case updatedAt = "updated_at"
    }
}

struct UserPreferenceContext: Codable {
    var id: String
    var fields: [String]
    var sourcePreferences: [String]
    var sessionTimeBudgetMin: Int
    var language: String
    var updatedAt: Date

    enum CodingKeys: String, CodingKey {
        case id
        case fields
        case sourcePreferences = "source_preferences"
        case sessionTimeBudgetMin = "session_time_budget_min"
        case language
        case updatedAt = "updated_at"
    }
}

struct UserMaterialContext: Codable, Identifiable {
    var id: String
    var title: String
    var sourceType: String
    var summary: String
    var url: URL?
    var filePath: String
    var authorsOrOwner: [String]
    var publishedDate: String
    var tags: [String]
    var relatedPlan: String
    var whySelected: String
    var fetchedAt: Date

    enum CodingKeys: String, CodingKey {
        case id
        case title
        case sourceType = "source_type"
        case summary
        case url
        case filePath = "file_path"
        case authorsOrOwner = "authors_or_owner"
        case publishedDate = "published_date"
        case tags
        case relatedPlan = "related_plan"
        case whySelected = "why_selected"
        case fetchedAt = "fetched_at"
    }
}

struct DirectionProfileSuggestionRequest: Codable {
    let longTermGoal: String
    let currentDirection: String
    let currentStage: String
    let backgroundSummary: String
    let targetCycle: String
    let timeBudgetMin: Int
    let fullCyclePlan: [String]
    let weeklyFocus: String
    let activeTasks: [String]
    let nextAction: String

    enum CodingKeys: String, CodingKey {
        case longTermGoal = "long_term_goal"
        case currentDirection = "current_direction"
        case currentStage = "current_stage"
        case backgroundSummary = "background_summary"
        case targetCycle = "target_cycle"
        case timeBudgetMin = "time_budget_min"
        case fullCyclePlan = "full_cycle_plan"
        case weeklyFocus = "weekly_focus"
        case activeTasks = "active_tasks"
        case nextAction = "next_action"
    }
}

struct DirectionProfileSuggestion: Codable {
    let fullCyclePlan: [String]
    let weeklyFocus: String
    let nextAction: String
    let activeTasks: [String]
    let trackingKeywords: [String]
    let fields: [String]
    let sourcePreferences: [String]
    let constraints: [String]
    let generationMode: String?

    enum CodingKeys: String, CodingKey {
        case fullCyclePlan = "full_cycle_plan"
        case weeklyFocus = "weekly_focus"
        case nextAction = "next_action"
        case activeTasks = "active_tasks"
        case trackingKeywords = "tracking_keywords"
        case fields
        case sourcePreferences = "source_preferences"
        case constraints
        case generationMode = "generation_mode"
    }
}

File: backend/app/schemas/user_context.py
from datetime import datetime
from enum import Enum
from typing import List, Optional

from pydantic import BaseModel, HttpUrl


class MaterialSourceType(str, Enum):
    pdf = "pdf"
    url = "url"
    manual = "manual"
    public_source = "public_source"
    arxiv = "arxiv"
    github = "github"
    official_doc = "official_doc"


class PersonalProfile(BaseModel):
    id: str = "default_profile"
    display_name: str = "新用户"
    goal: str = ""
    background_summary: str = ""
    current_stage: str = ""
    constraints: List[str] = []
    updated_at: datetime


class WorkLearningPlan(BaseModel):
    id: str = "active_plan"
    long_term_goal: str = ""
    target_cycle: str = ""
    full_cycle_plan: List[str] = []
    weekly_focus: str = ""
    active_tasks: List[str] = []
    next_action: str = ""
    tracking_keywords: List[str] = []
    updated_at: datetime


class UserPreference(BaseModel):
    id: str = "default_preferences"
    fields: List[str] = []
    source_preferences: List[MaterialSourceType] = [
        MaterialSourceType.arxiv,
        MaterialSourceType.official_doc,
        MaterialSourceType.url,
    ]
    session_time_budget_min: int = 30
    language: str = "zh-CN"
    updated_at: datetime


class UserMaterial(BaseModel):
    id: str
    title: str
    source_type: MaterialSourceType
    summary: str
    url: Optional[HttpUrl] = None
    file_path: str = ""
    authors_or_owner: List[str] = []
    published_date: str = ""
    tags: List[str] = []
    related_plan: str = ""
    why_selected: str = ""
    fetched_at: datetime


class UserMaterialCreate(BaseModel):
    title: str
    source_type: MaterialSourceType = MaterialSourceType.manual
    summary: str
    url: Optional[HttpUrl] = None
    file_path: str = ""
    authors_or_owner: List[str] = []
    published_date: str = ""
    tags: List[str] = []
    related_plan: str = ""


class UserContext(BaseModel):
    profile: PersonalProfile
    plan: WorkLearningPlan
    preferences: UserPreference
    materials: List[UserMaterial] = []


class DirectionProfileSuggestionRequest(BaseModel):
    long_term_goal: str = ""
    current_direction: str
    current_stage: str = ""
    background_summary: str = ""
    target_cycle: str = ""
    time_budget_min: int = 30
    full_cycle_plan: List[str] = []
    weekly_focus: str = ""
    active_tasks: List[str] = []
    next_action: str = ""


class DirectionProfileSuggestion(BaseModel):
    full_cycle_plan: List[str] = []
    weekly_focus: str
    next_action: str
    active_tasks: List[str] = []
    tracking_keywords: List[str] = []
    fields: List[str] = []
    source_preferences: List[MaterialSourceType] = []
    constraints: List[str] = []
    generation_mode: str = "fallback"


class RecommendationContext(BaseModel):
    derived_from: List[str] = []
    tracking_keywords: List[str] = []
    related_plan: str = ""
    user_preference: str = ""
    why_this_material_now: str = ""
